src/peft/tuners/adaption_prompt/layer.py

```diff
@@ -100,6 +104,15 @@ def forward(self, **kwargs):
         query_states = compute_query_states(model=self.model, **kwargs)
 
         previous_dtype = query_states.dtype
+
+        # Reshape and average the extra tensors
```

Comment:
No need to reshape and avg query states as the above key shape is (bsz, adapter_seq_len, num_kv_heads, head_dim), the value shape is (bsz, adapter_seq_len, num_kv_heads, head_dim) and query shape is (bsz, adapter_seq_len, num_heads, head_dim). Now, you would need to repeat the `num_kv_heads` to match `num_heads` as done in https://github.com/huggingface/transformers/blob/1c31b7aa3bb4e7ef24c77596d2a76f45a770159f/src/transformers/models/mistral/modeling_mistral.py#L193. After that the attn computation is same as normal MHA case.

Reply:
Thanks a lot, so rather than repeating the adapter output, I should repeat adapter_k and adapter_v.
`
        adapter_k = torch.repeat_interleave(
            adapter_k, repeats=factor, dim=1
        ) 
        adapter_v = torch.repeat_interleave(
            adapter_v, repeats=factor, dim=1
        )
`
Does this makes sense?